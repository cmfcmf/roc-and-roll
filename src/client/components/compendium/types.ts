import * as t from "typanion";
import { assert, IsExact } from "conditional-type-checks";
import { Opaque } from "type-fest";
import { isRRID } from "../../../shared/validation";

export type CompendiumTextEntry =
  | string
  | { type: "entries"; name: string; entries: CompendiumTextEntry[] }
  | { type: "list"; items: CompendiumTextEntry[] }
  | {
      type: "table";
      caption?: string;
      colLabels: string[];
      colStyles: unknown[];
      rows: CompendiumTextEntry[][];
    }
  | {
      type: "cell";
      roll: { min: number; max: number; pad?: boolean } | { exact: number };
    };

const __isTextEntryRecursive = t.makeValidator({
  test: (value, state): value is CompendiumTextEntry =>
    isTextEntry(value, state),
});

export const isTextEntry = t.isOneOf([
  t.isString(),
  t.isObject({
    type: t.isLiteral("entries"),
    name: t.isString(),
    entries: t.isArray(__isTextEntryRecursive),
  }),
  t.isObject({
    type: t.isLiteral("list"),
    items: t.isArray(__isTextEntryRecursive),
  }),
  t.isObject({
    type: t.isLiteral("table"),
    caption: t.isOptional(t.isString()),
    colLabels: t.isArray(t.isString()),
    colStyles: t.isArray(t.isUnknown()),
    rows: t.isArray(t.isArray(__isTextEntryRecursive)),
  }),
  t.isObject({
    type: t.isLiteral("cell"),
    roll: t.isOneOf(
      [
        t.isObject({
          exact: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
        }),
        t.isObject({
          min: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
          max: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
          pad: t.isOptional(t.isBoolean()),
        }),
      ],
      { exclusive: true }
    ),
  }),
]);

// Make sure that the schema really matches the TextEntry type.
assert<IsExact<t.InferType<typeof isTextEntry>, CompendiumTextEntry>>(true);

export type CompendiumSpell = {
  components: {
    v?: boolean;
    s?: boolean;
    m?:
      | string
      | { text: string; cost?: number; consume?: boolean | "optional" };
  };
  duration: Array<
    | { type: "instant" | "special" }
    | { type: "permanent"; ends: Array<"dispel" | "trigger"> }
    | {
        type: "timed";
        duration: {
          type: "round" | "minute" | "hour" | "day";
          amount: number;
          upTo?: boolean;
        };
        concentration?: boolean;
      }
  >;
  level: number;
  name: string;
  range:
    | {
        type:
          | "point"
          | "radius"
          | "line"
          | "cone"
          | "hemisphere"
          | "sphere"
          | "cube";
        distance: {
          type: "feet" | "miles";
          amount: number;
        };
      }
    | {
        type: "point";
        distance: {
          type: "self" | "touch" | "sight" | "unlimited";
        };
      }
    | { type: "special" };
  source: string;
  page: number;
  time: Array<
    | { number: number; unit: "action" | "bonus" | "minute" | "hour" }
    | { number: number; unit: "reaction"; condition: string }
  >;
  entries: CompendiumTextEntry[];
  entriesHigherLevel?: CompendiumTextEntry[];
  srd?: boolean | string;
};

export const isSpell = t.isObject(
  {
    components: t.isObject({
      v: t.isOptional(t.isBoolean()),
      s: t.isOptional(t.isBoolean()),
      m: t.isOptional(
        t.isOneOf(
          [
            t.isString(),
            t.isObject({
              text: t.isString(),
              cost: t.isOptional(
                t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()])
              ),
              consume: t.isOptional(
                t.isOneOf([t.isBoolean(), t.isLiteral("optional")], {
                  exclusive: true,
                })
              ),
            }),
          ],
          { exclusive: true }
        )
      ),
    }),
    duration: t.isArray(
      t.isOneOf(
        [
          t.isObject({
            type: t.isEnum(["instant", "special"] as const),
          }),
          t.isObject({
            type: t.isLiteral("permanent"),
            ends: t.isArray(
              t.isOneOf([t.isLiteral("dispel"), t.isLiteral("trigger")], {
                exclusive: true,
              })
            ),
          }),
          t.isObject({
            type: t.isLiteral("timed"),
            duration: t.isObject({
              type: t.isEnum(["round", "minute", "hour", "day"] as const),
              amount: t.applyCascade(t.isNumber(), [
                t.isInteger(),
                t.isPositive(),
              ]),
              upTo: t.isOptional(t.isBoolean()),
            }),
            concentration: t.isOptional(t.isBoolean()),
          }),
        ],
        { exclusive: true }
      )
    ),
    level: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
    name: t.isString(),
    range: t.isOneOf(
      [
        t.isObject({
          type: t.isEnum([
            "point",
            "radius",
            "line",
            "cone",
            "hemisphere",
            "sphere",
            "cube",
          ] as const),
          distance: t.isObject({
            type: t.isEnum(["feet", "miles"] as const),
            amount: t.applyCascade(t.isNumber(), [
              t.isInteger(),
              t.isPositive(),
            ]),
          }),
        }),
        t.isObject({
          type: t.isLiteral("point"),
          distance: t.isObject({
            type: t.isEnum(["self", "touch", "sight", "unlimited"] as const),
          }),
        }),
        t.isObject({ type: t.isLiteral("special") }),
      ],
      {
        exclusive: true,
      }
    ),
    source: t.isString(),
    page: t.applyCascade(t.isNumber(), [t.isInteger(), t.isPositive()]),
    time: t.isArray(
      t.isOneOf(
        [
          t.isObject({
            number: t.applyCascade(t.isNumber(), [
              t.isInteger(),
              t.isPositive(),
            ]),
            unit: t.isEnum(["action", "bonus", "minute", "hour"] as const),
          }),
          t.isObject({
            number: t.applyCascade(t.isNumber(), [
              t.isInteger(),
              t.isPositive(),
            ]),
            unit: t.isLiteral("reaction"),
            condition: t.isString(),
          }),
        ],
        { exclusive: true }
      )
    ),
    entries: t.isArray(isTextEntry),
    entriesHigherLevel: t.isOptional(t.isArray(isTextEntry)),
    srd: t.isOptional(
      // If this is a string, then the spell has a different name in the SRD.
      t.isOneOf([t.isBoolean(), t.isString()], { exclusive: true })
    ),
  },
  // There are a lot more properties, which we currently ignore.
  { extra: t.isUnknown() }
);

// Make sure that the schema really matches the Spell type.
assert<IsExact<t.InferType<typeof isSpell>, CompendiumSpell>>(true);

export const isCompendiumData = t.isObject({
  spell: t.isArray(isSpell),
});

export type CompendiumData = {
  spell: CompendiumSpell[];
};

// Make sure that the schema really matches the CompendiumData type.
assert<IsExact<t.InferType<typeof isCompendiumData>, CompendiumData>>(true);

export type CompendiumSourceID = Opaque<string, "compendiumSource">;

export type CompendiumSource = {
  id: CompendiumSourceID;
  title: string;
  meta: string;
  data: CompendiumData;
};

export const isCompendiumSource = t.isObject({
  id: isRRID<CompendiumSourceID>(),
  title: t.isString(),
  meta: t.isString(),
  data: isCompendiumData,
});

// Make sure that the schema really matches the TextEntry type.
assert<IsExact<t.InferType<typeof isCompendiumSource>, CompendiumSource>>(true);
