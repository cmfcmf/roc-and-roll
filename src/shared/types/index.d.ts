declare const __VERSION__: string;

declare const IMPORT_META_URL: string;

declare module "!!raw-loader!*" {
  const contents: string;
  export = contents;
}

declare module "*.css" {
  const classNames: Record<string, string>;
  export = classNames;
}

declare module "*.scss" {
  const classNames: Record<string, string>;
  export = classNames;
}

declare module "*.mp3" {
  const mp3: string;
  export = mp3;
}

declare module "*.svg" {
  const svg: string;
  export = svg;
}

declare module "*.png" {
  const png: string;
  export = png;
}

declare module "*.jpg" {
  const jpg: string;
  export = jpg;
}

declare module "*.jpeg" {
  const jpeg: string;
  export = jpeg;
}

declare module "*.glb" {
  const glb: string;
  export = glb;
}

declare namespace NodeJS {
  interface ProcessEnv {
    readonly NODE_ENV: "development" | "production" | "test";
    readonly HEROKU?: string;
  }
}

interface Window {
  MODDING?: unknown;
}

// Make JSON and isArray checks return unknown instead of any.
//
// https://github.com/Microsoft/TypeScript/issues/26188#issuecomment-489659669
interface JSON {
  parse(
    text: string,
    reviver?: (this: unknown, key: string, value: unknown) => unknown
  ): unknown;

  stringify(
    text: unknown,
    replacer?: (this: unknown, key: string, value: unknown) => unknown,
    space?: string | number
  ): string;
}

interface Body {
  json(): Promise<unknown>;
}

declare module "react-lag-radar" {
  const component: React.ComponentType<{
    frames?: number;
    speed?: number;
    size?: number;
    inset?: number;
  }>;

  export = component;
}
