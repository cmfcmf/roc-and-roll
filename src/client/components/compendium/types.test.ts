import { isTextEntry } from "./types";

describe("text entry schema", () => {
  it.each([
    ["test", true],
    [null, false],
    [{ type: "cell", roll: { exact: 3 } }, true],
    [{ type: "cell", roll: { min: 2, max: 3 } }, true],
    [{ type: "cell", roll: { min: 2, max: 3, pad: true } }, true],
    [{ type: "list", items: ["foo", { type: "list", items: [] }] }, true],
  ])("works", (entry, result) => {
    expect(isTextEntry.safeParse(entry).success).toBe(result);
  });

  it("uses correct keys for recursive errors", () => {
    expect(
      isTextEntry.safeParse({
        type: "list",
        items: ["foo", "bar"],
      }).success
    ).toBe(true);

    const validationResult = isTextEntry.safeParse({
      type: "list",
      items: ["foo", false],
    });
    if (validationResult.success) {
      fail();
    }

    expect(validationResult.error).toMatchSnapshot();
  });
});
