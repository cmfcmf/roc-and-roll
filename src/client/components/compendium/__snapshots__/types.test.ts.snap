// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text entry schema uses correct keys for recursive errors 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "object",
            "path": [],
            "message": "Expected string, received object"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "entries",
            "received": "list",
            "path": [
              "type"
            ],
            "message": "Expected entries, received list"
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "name"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              "entries"
            ],
            "message": "Required"
          },
          {
            "code": "unrecognized_keys",
            "keys": [
              "items"
            ],
            "path": [],
            "message": "Unrecognized key(s) in object: 'items'"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_union",
            "unionErrors": [
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "string",
                    "received": "boolean",
                    "path": [
                      "items",
                      1
                    ],
                    "message": "Expected string, received boolean"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "boolean",
                    "path": [
                      "items",
                      1
                    ],
                    "message": "Expected object, received boolean"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "boolean",
                    "path": [
                      "items",
                      1
                    ],
                    "message": "Expected object, received boolean"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "boolean",
                    "path": [
                      "items",
                      1
                    ],
                    "message": "Expected object, received boolean"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "boolean",
                    "path": [
                      "items",
                      1
                    ],
                    "message": "Expected object, received boolean"
                  }
                ],
                "name": "ZodError"
              }
            ],
            "path": [
              "items",
              1
            ],
            "message": "Invalid input"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "table",
            "received": "list",
            "path": [
              "type"
            ],
            "message": "Expected table, received list"
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              "colLabels"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              "colStyles"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "array",
            "received": "undefined",
            "path": [
              "rows"
            ],
            "message": "Required"
          },
          {
            "code": "unrecognized_keys",
            "keys": [
              "items"
            ],
            "path": [],
            "message": "Unrecognized key(s) in object: 'items'"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "cell",
            "received": "list",
            "path": [
              "type"
            ],
            "message": "Expected cell, received list"
          },
          {
            "code": "invalid_union",
            "unionErrors": [
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "undefined",
                    "path": [
                      "roll"
                    ],
                    "message": "Required"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "object",
                    "received": "undefined",
                    "path": [
                      "roll"
                    ],
                    "message": "Required"
                  }
                ],
                "name": "ZodError"
              }
            ],
            "path": [
              "roll"
            ],
            "message": "Invalid input"
          },
          {
            "code": "unrecognized_keys",
            "keys": [
              "items"
            ],
            "path": [],
            "message": "Unrecognized key(s) in object: 'items'"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;
